README.TXT
mypaedia-package --- FreePWING 版『マイペディア』用 lookup 補助パッケージ

    かずひこ <kazuhiko@ring.gr.jp>
    http://openlab.ring.gr.jp/edict/mypaedia-fpw/

このディレクトリにはマイペディアのマルチメディア・データを lookup で利用
するための elisp ファイルを収めています。

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
1. 必要なもの

    FreePWING 版マイペディア 1.3.2 以降
	http://openlab.ring.gr.jp/edict/mypaedia-fpw/
	作者の所有する『マイペディア for Win (2000年版)』、『マイペディ
	ア PC Success 版』で動作確認済。

    lookup 1.2 (おそらく 1.0 や 1.1 でも可)
	http://openlab.ring.gr.jp/lookup/

そのほか，JPEG，BMP，PNG，MIDI，WAV，AVI を表示・再生できるソフトウェア
が必要です (全部はなくても可)。

(参考 URL)

    TiMidity++ -- MIDI 再生
	http://www.goice.co.jp/member/mo/timidity/
    sox -- WAV 再生
	http://home.sprynet.com/~cbagwell/sox.html
    xanim -- AVI 再生 (cinepak 用の dll も必要)
	http://xanim.va.pubnix.com/home.html
    ImageMagick -- 各種画像の表示
	http://www.wizards.dupont.com/cristy/ImageMagick.html

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
2. 利用方法

~/.lookup で，

   (setq lookup-search-agents '((ndeb "/usr/share/dict/mypaedia")))
   (setq lookup-package-directory "/usr/share/dict/package")
   (lookup-use-package "ndeb+/usr/share/dict/mypaedia" "mypaedia-fpw")

   ; BH.DAT，CH.DAT，MAP.DAT，PICT.DAT のある場所
   (setq mypaedia-data-directory "/mnt/cdrom/INDEX")

   ; *.MID，*.WAV，*.AVI のある場所
   (setq mypaedia-av-directory "/mnt/cdrom/DATA")

   (setq mypaedia-play-midi-process "timidity")
   (setq mypaedia-play-wav-process "soxplay")
   (setq mypaedia-play-avi-process "xanim")
   (setq mypaedia-display-image-process "display")
   ; "perl のプロセス名。パスが通っていない場合はフルパスで記述すること。
   (setq perl-process "perl")

   ; 一時ファイルの置き場所。
   (setq temorary-file-directory "/tmp")

のように設定して，lookup-package-directory に mypaedia-fpw.el と
extract-mypaedia.pl をコピーして下さい。xemacs および emacs-21 では画像
のインライン表示に対応していますが，xemacs ではキャッシュが効いていない
ので，画像を再表示したいときには `g' でリロードする必要があります。なお，
インライン表示させたくないときは，~/.lookup で

   (setq mypaedia-image-inline nil)

としてください。また，

画像，MIDI，WAV，AVI のうち表示・再生したくないものについては，
~/.lookup でそれぞれ

   (setq mypaedia-display-image-process nil)
   (setq mypaedia-play-midi-process nil)
   (setq mypaedia-play-wav-process nil)
   (setq mypaedia-play-avi-process nil)

としてください。

音声や動画を，検索と同時に再生したい場合は，~/.lookup でそれぞれ

   (setq mypaedia-sound-without-notice t)
   (setq mypaedia-video-without-notice t)

としてください。

なお lookup で検索する際に、データ量の多いエントリで本文が途中で <more
point=5> となって途切れる場合は、lookup-max-text の値を変更してください。
詳しくは lookup 付属のドキュメントをご覧ください。

> lookup-max-text : 検索時に表示するエントリ本文の最大長。0 を指定すると、
> 全文を表示する。

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
2. ライセンス

mypaedia-package はフリーソフトウェアです。あなたは，Free Software
Foundation が公表した GNU General Public License (GNU 一般公有使用許諾)
バージョン 2 あるいはそれ以降の各バージョンの中からいずれかを選択し，そ
のバージョンが定める条項に従って本プログラムを再頒布または変更することが
できます。

mypaedia-package は有用とは思いますが，頒布にあたっては，市場性及び特定
目的適合性についての暗黙の保証を含めて，いかなる保証も行いません。詳細に
ついては GNU General Public License をお読み下さい。

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
3. 謝辞

原型を作ってくださった西田圭介さん，大幅に拡張してくださった是枝和義さん，
動作確認などにご協力くださった野宮賢さん、OKUNISHI さん、そして lookup
Mailing List *1 のみなさんに感謝します。

 *1 http://openlab.ring.gr.jp/lookup/


ChangeLog
2001-12-01  Kazuhiko  <kazuhiko@ring.gr.jp>

	* mypaedia-fpw-package-1.0.7 公開
	* mypaedia-fpw.el (mypaedia-arrange-images): Emacs-21 に対応。

2001-06-24  Kazuhiko  <kazuhiko@ring.gr.jp>

	* mypaedia-fpw-package-1.0.6 公開
	* mypaedia-fpw.el (lookup-package-dictionary-options-alist): 辞書
	名が大文字でも小文字でも対応するように変更。

2000-10-06 Kazuhiko Shiozaki <kazuhiko@ring.gr.jp>

	* mypaedia-fpw-package-1.0.4 公開
	* mypaedia-fpw.el XEmacs でのインライン表示を、一時的に
	  format-alist を変更することにより XEmacs-21.2.x に対応。

2000-08-05 Kazuhiko Shiozaki <kazuhiko@ring.gr.jp>

	* mypaedia-fpw.el display-image-inline から mypaedia-image-inline に
	  改名。

2000-08-01 Kazuhiko Shiozaki <kazuhiko@ring.gr.jp>

	* mypaedia-fpw-package-1.0.3 公開
	* extract-mypaedia.pl extract.pl を改名。

2000-06-17 Kazuhiko Shiozaki <kazuhiko@ring.gr.jp>

	* mypaedia-fpw-package1.0.2 公開（mypaedia-package を改称）。

2000-06-16 Kazuhiko Shiozaki <kazuhiko@ring.gr.jp>

	* mypaedia-fpw.el mypaedia.el を改称。
	* extract.pl 新設。
	* mypaedia-fpw.el ファイルの切出しを elisp ではなく perl で実装した。
	  これにより、emacs の int の制限を回避。

2000-04-12 Kazuhiko Shiozaki <kazuhiko@ring.gr.jp>

	* mypaedia-package1.0.1 公開

	* mypaedia.el 音声や動画のプロセスを mini buffer に表示するように
	  した。そのほか細かなバグ修正。
	* README.TXT lookup-max-text に関する記述などを追加。

2000-04-11 Kazuhiko Shiozaki <kazuhiko@ring.gr.jp>

	* mypaedia-package1.0 公開

	* mypaedia.el xemacs での非インライン表示に対応。WAV, AVI のいきな
	  り再生に対応。

2000-04-10 Kazuyoshi Koreeda <k_koreed@d2.dion.ne.jp>

	* mypaedia.el MIDI のいきなり再生に対応。ファイル名タグを非表示に。

2000-03-31 Kazuyoshi Koreeda <k_koreed@d2.dion.ne.jp>

	* mypaedia.el 外部プロセスを用いた画像表示に対応。

2000-03-26 Kazuhiko Shiozaki <kazuhiko@ring.gr.jp>

	* mypaedia.el MIDI，WAV，AVI のファイル名を表示するようにした。

2000-03-17 Kazuhiko Shiozaki <kazuhiko@ring.gr.jp>
	* mypaedia.el 外部プロセスを用いた MIDI，WAV，AVI の再生に対応。

2000-03-11 Keisuke Nishida <kxn30@po.cwru.edu>

	* mypaedia.el xemacs での画像のインライン表示に対応。


extract-mypaedia.pl
#                                                         -*- Perl -*-
# Copyright (C) 2000  Kazuhiko Shiozaki
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2, or (at your option)
# any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#

require 5.005;

use strict;
use warnings;
use English;
use FileHandle;

my $file = shift;
my $offset = hex shift;
my $size = hex shift;
my $tmp_file = shift;

# print "File:$file\n";
# print "Offset:$offset\n";
# print "Size:$size\n";
# print "Tmp_File:$tmp_file\n";

my $handle_in = FileHandle->new();
if (!$handle_in->open("$file", 'r')) {
  die "$PROGRAM_NAME: failed to open the file, $ERRNO: $file\n";
}
binmode $handle_in;

my $handle_out = FileHandle->new();
if (!$handle_out->open("$tmp_file", 'w')) {
  die "$PROGRAM_NAME: failed to open the file, $ERRNO: $tmp_file\n";
}
binmode $handle_out;

my $content;
seek($handle_in, $offset, 0) || die "Seek Error\n";
read $handle_in, $content, $size || die "Read Error\n";
print $handle_out $content || die "Write Error\n";


mypaedia-fpw.el
;;; mypaedia-fpw.el --- supplement file for 『マイペディア９９』
;; $Id: mypaedia-fpw.el,v 1.8 2001/11/25 05:04:49 kazuhiko Exp $
;;
;; Copyright (C) 2000 Keisuke Nishida <kxn30@po.cwru.edu>
;; Copyright (C) 2000 Kazuhiko Shiozaki <kazuhiko@ring.gr.jp>
;; Copyright (C) 2000 Kazuyoshi KOREEDA <k_koreed@d2.dion.ne.jp>

;; This program is free software; you can redistribute it and/or
;; modify it under the terms of the GNU General Public License
;; as published by the Free Software Foundation; either version 2
;; of the License, or (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program; if not, write to the Free Software Foundation,
;; Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

;;; Code:

(require 'lookup-package)
(require 'lookup-content)
(require 'ndeb)

(defvar mypaedia-data-directory "/usr/share/dict/epwing/MYPAEDIA_DATA/INDEX"
  "BH.DAT, CH.DAT, MAP.DAT, PICT.DAT のある場所。")
(defvar mypaedia-av-directory "/usr/share/dict/epwing/MYPAEDIA_DATA/DATA"
  "*.MID, *.WAV, *.AVI のある場所。")
(defvar mypaedia-play-midi-process "timidity"
  "MIDI を再生するプロセス名。nil なら再生しない。")
(defvar mypaedia-play-wav-process "alsaplayer"
  "WAV を再生するプロセス名。nil なら再生しない。")
(defvar mypaedia-play-avi-process "vlc"
  "AVI を再生するプロセス名。nil なら再生しない。")
(defvar mypaedia-display-image-process "display"
  "画像 を表示するプロセス名。nil なら表示しない。")
(defvar mypaedia-sound-without-notice nil
  "t なら検索と同時に音声を再生する。")
(defvar mypaedia-video-without-notice nil
  "t なら検索と同時に動画を再生する。")
(defvar mypaedia-image-inline t
  "nil なら (可能な場合でも) 画像をインライン表示しない。")
(defvar perl-process "/usr/bin/perl"
  "perl のプロセス名。パスが通っていない場合はフルパスで記述すること。")

(setq ndeb-max-image-size 4194304)

(defconst mypaedia-arranges
  '(lookup-arrange-gaijis
    mypaedia-arrange-midi
    mypaedia-arrange-wav
    mypaedia-arrange-avi
    mypaedia-arrange-images
    lookup-arrange-references
    lookup-arrange-default-headings
    lookup-arrange-fill-lines))

(defvar mypaedia-dictionary-options
  (list (cons ':arranges mypaedia-arranges)))

(defvar mypaedia-process-file-alist '())

;;
;; 元からある lookup-content-follow-linkの拡張
;;
(unless (fboundp 'lookup-content-follow-link:old)
  (fset 'lookup-content-follow-link:old
	(symbol-function 'lookup-content-follow-link))
  (defun lookup-content-follow-link ()
    (interactive)
    (let ((action (get-text-property (point) 'action)))
      (if action
	  (funcall action (point))
	(lookup-content-follow-link:old)))))

(defun mypaedia-string-to-number (string)
  (let ((len (length string))
	(number 0) (i 0) c)
    (while (< i len)
      (setq number (* number 16))
      (setq c (aref string i))
      (cond
       ((and (<= ?0 c) (<= c ?9)) (setq number (+ number (- c ?0))))
       ((and (<= ?a c) (<= c ?f)) (setq number (+ number (- c ?a -10))))
       ((and (<= ?A c) (<= c ?F)) (setq number (+ number (- c ?A -10))))
       (t (setq i len)))
      (setq i (1+ i)))
    number))

(defun mypaedia-arrange-images (entry)
  (while (re-search-forward
	  "<image=\\([^:>]+\\):\\([^:]+\\):\\([^>]+\\)>" nil t)
    (let ((file (concat (match-string 1) ".DAT"))
	  (offset (match-string 2))
	  (length (match-string 3))
	  (start (match-beginning 0))
	  (end (match-end 0)))
	  ;; Find data file.
      (if (file-exists-p (expand-file-name file mypaedia-data-directory))
	  (setq file (expand-file-name file mypaedia-data-directory)))
      (if mypaedia-image-inline
	  (cond ((featurep 'xemacs)
		 (progn
		   (replace-match "\n")
		   (let ((offset (mypaedia-string-to-number offset))
			 (length (mypaedia-string-to-number length))
			 (format-alist
			  '((image/jpeg "JPEG image" "\377\330\377\340\000\020JFIF"
					image-decode-jpeg nil t image-mode)
			    (image/png "Portable Network Graphics" "\211PNG"
				       image-decode-png nil t image-mode))))
		     (insert-file-contents-internal file nil offset
						    (+ offset length)))))
		((featurep 'image)
		 (progn
		   (replace-match "\n")
		   (let ((tmp-img-file
			  (make-temp-name
			   (expand-file-name "mp" temporary-file-directory))))
		     (call-process
		      perl-process nil nil nil
		      (expand-file-name "extract-mypaedia.pl"
					lookup-package-directory)
		      file offset length tmp-img-file)
		     (let ((glyph
			    (with-temp-buffer
			      (insert-file-contents-literally
			       tmp-img-file  nil 0 ndeb-max-image-size)
			      (string-make-unibyte
			       (buffer-substring-no-properties
				(point-min) (point-max))))))
		       (insert-image
			(create-image glyph nil t :ascent 'center)))
		       (delete-file tmp-img-file)
		       ))))
	(replace-match "→[画像]")
	(add-text-properties start
			     (+ (length "→[画像]") start)
			     (list 'action 'mypaedia-display-image
				   'file  file
				   'offset offset
				   'mouse-face 'highlight
				   'face 'lookup-reference-face
				   'length   length))))))
;;
;; 外部プロセスを利用したイメージの表示
;;
(defun mypaedia-display-image (pos)
  (let* ((file (get-text-property pos 'file))
	 (offset (get-text-property pos 'offset))
	 (length (get-text-property pos 'length))
	 (tmp-img-file (make-temp-name
			(expand-file-name "mp" temporary-file-directory))))
    (if mypaedia-display-image-process
	(progn
	  (call-process perl-process nil nil nil
			(expand-file-name "extract-mypaedia.pl"
					  lookup-package-directory)
			file offset length tmp-img-file)
	  (mypaedia-start-process
	   mypaedia-display-image-process nil tmp-img-file t)))))
;;
;; 外部プロセスの呼出し
;;
(defun mypaedia-start-process (program options file &optional delete-file)
    (message "Starting %s ..." program)
    (let ((pro (apply (function start-process)
		      (format "*mypaedia %s*" program)
		      nil
		      program
		      (append options (list file)))))
      	(message "Starting %s ... done" program)
	(set-process-sentinel pro 'mypaedia-start-process-sentinel)
	(setq mypaedia-process-file-alist
		(cons (cons pro file)
		      (if delete-file
			  mypaedia-process-file-alist
			nil)))))
;;
;; プロセスの状態が変更されたときにファイルを削除する。
;;
(defun mypaedia-start-process-sentinel (process event)
  (let ((al (assoc process mypaedia-process-file-alist)))
    (and (cdr al) (delete-file (cdr al)))
    (setq mypaedia-process-file-alist
	  (delete al mypaedia-process-file-alist))))

;;
;; 外部プロセスを利用した MIDI の再生
;;
(defun mypaedia-arrange-midi (entry)
  (while (re-search-forward "<sound=\\([^>]+mid\\)>" nil t)
    (let ((file (upcase (match-string 1)))
	  (start (match-beginning 0))
	  (end (match-end 0)))
      (if mypaedia-sound-without-notice
	  (mypaedia-play-midi-sub file t))
      (replace-match "→[音声]")
      (add-text-properties start
			   (+ (length "→[音声]") start)
			   (list 'action 'mypaedia-play-midi
				 'file file
				 'mouse-face 'highlight
				 'face 'lookup-reference-face)))))

(defun mypaedia-play-midi-sub (file &optional bulk)
  (if mypaedia-play-midi-process
      (let ((midi-file (expand-file-name file mypaedia-av-directory)))
	(if (not (file-exists-p midi-file))
	    (or bulk (error "%s: not found." midi-file))
	  (if (get-process "lookup-play-midi")
	      (delete-process "lookup-play-midi")
	    (if (get-process "lookup-play-wav")
		(delete-process "lookup-play-wav")))
	  (message "Starting %s ..." mypaedia-play-midi-process)
	  (start-process "lookup-play-midi" nil
			 mypaedia-play-midi-process midi-file)
	  (message "Starting %s ... done" mypaedia-play-midi-process)))))

(defun mypaedia-play-midi (pos)
  (mypaedia-play-midi-sub (get-text-property pos 'file)))

;;
;; 外部プロセスを利用した WAV の再生
;;
(defun mypaedia-arrange-wav (entry)
  (while (re-search-forward "<sound=\\([^>]+wav\\)>" nil t)
    (let ((file (upcase (match-string 1)))
	  (start (match-beginning 0))
	  (end (match-end 0)))
      (if mypaedia-sound-without-notice
	  (mypaedia-play-wav-sub file t))
      (replace-match "→[音声]")
      (add-text-properties start
			   (+ (length "→[音声]") start)
			   (list 'action 'mypaedia-play-wav
				 'file file
				 'mouse-face 'highlight
				 'face 'lookup-reference-face)))))

(defun mypaedia-play-wav-sub (file &optional bulk)
  (if mypaedia-play-wav-process
      (let ((wav-file (expand-file-name file mypaedia-av-directory)))
	(if (not (file-exists-p wav-file))
	    (or bulk (error "%s: not found." wav-file))
	  (if (get-process "lookup-play-midi")
	      (delete-process "lookup-play-midi")
	    (if (get-process "lookup-play-wav")
		(delete-process "lookup-play-wav")))
	  (message "Starting %s ..." mypaedia-play-wav-process)
	  (start-process "lookup-play-wav" nil
			 mypaedia-play-wav-process wav-file)
	  (message "Starting %s ... done" mypaedia-play-wav-process)))))

(defun mypaedia-play-wav (pos)
  (mypaedia-play-wav-sub (get-text-property pos 'file)))

;;
;; 外部プロセスを利用した AVI の再生
;;
(defun mypaedia-arrange-avi (entry)
  (while (re-search-forward "<video=\\([^>]+avi\\)>" nil t)
    (let ((file (upcase (match-string 1)))
	  (start (match-beginning 0))
	  (end (match-end 0)))
      (if mypaedia-video-without-notice
	  (mypaedia-play-avi-sub file t))
      (replace-match "→[動画]")
      (add-text-properties start
			   (+ (length "→[動画]") start)
			   (list 'action 'mypaedia-play-avi
				 'file  file
				 'mouse-face 'highlight
				 'face 'lookup-reference-face)))))

(defun mypaedia-play-avi-sub (file &optional bulk)
  (if mypaedia-play-avi-process
      (let ((avi-file (expand-file-name file mypaedia-av-directory)))
	(if (not (file-exists-p avi-file))
	    (or bulk (error "%s: not found." avi-file))
	  (if (get-process "lookup-play-avi")
	      (delete-process "lookup-play-avi"))
	  (message "Starting %s ..." mypaedia-play-avi-process)
	  (start-process "lookup-play-avi" nil
			 mypaedia-play-avi-process avi-file)
	  (message "Starting %s ... done" mypaedia-play-avi-process)))))

(defun mypaedia-play-avi (pos)
  (mypaedia-play-avi-sub (get-text-property pos 'file)))

(setq lookup-package-dictionary-options-alist
      (list (cons (cond ((eq lookup-package-agent 'ndtp) "mypaedia/mypaedia")
			((eq lookup-package-agent 'ndtp) "MYPAEDIA/MYPAEDIA")
			((eq lookup-package-agent 'ndeb) "mypaedia")
			((eq lookup-package-agent 'ndeb) "MYPAEDIA"))
		  mypaedia-dictionary-options)))

(provide 'mypaedia-fpw)

;;; mypaedia-fpw.el ends here
